src/main/java/com/iota/iri/LedgerValidator.java

```diff
@@ -198,103 +187,85 @@ private void updateConsistentHashes(final Snapshot snapshot, Hash tip, int index
      * @throws Exception
      */
     protected void init() throws Exception {
-        MilestoneViewModel latestConsistentMilestone = buildSnapshot(milestone.latestSnapshot);
+        MilestoneViewModel latestConsistentMilestone = buildSnapshot();
         if(latestConsistentMilestone != null) {
             milestone.latestSolidSubtangleMilestone = latestConsistentMilestone.getHash();
             milestone.latestSolidSubtangleMilestoneIndex = latestConsistentMilestone.index();
         }
     }
 
-    public static boolean isApproved(Snapshot snapshot, Hash hash) {
-        synchronized (snapshot.approvalsSyncObject) {
-            return snapshot.approvedHashes.contains(hash);
-        }
-    }
-
-
     /**
      * Only called once upon initialization, this builds the {latestSnapshot} state up to the most recent
      * solid milestone confirmed. It gets the earliest confirmed, and while checking for consistency, patches the next
      * newest confirmed diff into its map.
      * @return              the most recent consistent milestone with a confirmed.
      * @throws Exception
      */
-    private MilestoneViewModel buildSnapshot(Snapshot latestSnapshot) throws Exception {
+    private MilestoneViewModel buildSnapshot() throws Exception {
         MilestoneViewModel consistentMilestone = null;
-        synchronized (latestSnapshot.snapshotSyncObject) {
-            Snapshot updatedSnapshot = latestSnapshot.patch(new HashMap<>(), 0);
-            StateDiffViewModel stateDiffViewModel;
-            MilestoneViewModel snapshotMilestone = MilestoneViewModel.firstWithSnapshot(tangle);
-            while (snapshotMilestone != null) {
-                stateDiffViewModel = StateDiffViewModel.load(tangle, snapshotMilestone.getHash());
-                updatedSnapshot = updatedSnapshot.patch(stateDiffViewModel.getDiff(), snapshotMilestone.index());
-                if (updatedSnapshot.isConsistent()) {
-                    consistentMilestone = snapshotMilestone;
-                    latestSnapshot.merge(updatedSnapshot);
-                    snapshotMilestone = snapshotMilestone.nextWithSnapshot(tangle);
-                }
+        StateDiffViewModel stateDiffViewModel;
+        milestone.latestSnapshot.rwlock.writeLock().lock();
+        MilestoneViewModel snapshotMilestone = MilestoneViewModel.firstWithSnapshot(tangle);
+        while (snapshotMilestone != null) {
+            stateDiffViewModel = StateDiffViewModel.load(tangle, snapshotMilestone.getHash());
+            if(Snapshot.isConsistent(milestone.latestSnapshot.patch(stateDiffViewModel.getDiff()))) {
+                milestone.latestSnapshot.merge(stateDiffViewModel.getDiff(), snapshotMilestone.index());
+                consistentMilestone = snapshotMilestone;
+                snapshotMilestone = snapshotMilestone.nextWithSnapshot(tangle);
             }
         }
+        milestone.latestSnapshot.rwlock.writeLock().unlock();
```

Comment:
Should these unlocks be in finally blocks? If an exception is thrown during lockings, there will be deadlock.

Reply:
Thanks